First cut at addressing per hostname override

kodster28 opened the pull request:
Meant to address https://jira.cfops.it/browse/PCX-838 submitted by @

Made small changes to other files, but here's a screenshot of the main update to the **Monitors** page for reviewing convenience.

![image](https://user-images.githubusercontent.com/26727299/108546327-95523200-72ae-11eb-8631-3abf864b2e60.png)

bbatraskicf commented:
This overall looks good. I see we focus heavily though on the monitor side which looks good. I'd also like to highlight the table we make available for the monitors lending to easier workflow and having more relevant data in one view. 

I think we need to further highlight the impact of being able to use host header overrides on the origin allowing utilization of products such as google appengine, heroku, amazon s3, etc. We give good examples for monitors but we should do the same for the requests going to those origins and how we are able to properly route and support the addition of those third party applications. I think highlighting the origin host header, the request host header, how it would break if not changed on the origin when using those other services, and by allowing this very easy setup, we then are able to properly route eyeball requests correctly.

kodster28 commented:
Howdy @bbatraskicf , 

Pushed some changes to hopefully address your suggestions.

**1**. Updated the monitor page and included that screenshot of the list of origin overrides.
![image](https://user-images.githubusercontent.com/26727299/108904287-ee3e0500-75e3-11eb-81cd-2ef14e997540.png)

**2**. Added details about the origin host header and provided example use cases. If we need more details for those examples, you might have to give me the words for them.
![image](https://user-images.githubusercontent.com/26727299/108904429-16c5ff00-75e4-11eb-8a97-89af22f07737.png)

**3**. Included links to **2** in the Create API and Create via Dashboard steps.

Thoughts?

kodster28 commented:
@bbatraskicf , I updated this PR to reflect the changes y'all made to the host headers for security.

Am I using the right words here?

If you agree with these changes (and the overall PR), let me know! If not, happy to do some more word-smithing with your guidance.

![image](https://user-images.githubusercontent.com/26727299/112048646-a21bad00-8b1c-11eb-9e9c-87a5ab4a137a.png)

![image](https://user-images.githubusercontent.com/26727299/112048658-a5af3400-8b1c-11eb-9677-d2d5158c8c97.png)

pedrosousa commented on products/load-balancing/src/content/understand-basics/monitors.md:
```diff
@@ -257,6 +257,44 @@ The sub-string must appear within the first 10KiB of your response body.
+
+When your application needs specialized routing (CNAME setup or custom hosts like Heroku), change the `Host` header used in health checks.
+
+You can set these headers on a [specific origin](/understand-basics/pools#per-origin-host-header-override) or a monitor. Headers set on an origin always override headers set on a monitor.
```
Not sure if you need the "always" (and the same in the next section).

pedrosousa commented on products/load-balancing/src/content/understand-basics/monitors.md:
```diff
@@ -257,6 +257,44 @@ The sub-string must appear within the first 10KiB of your response body.
+
+  - Pool 1:
+
+    - Origin 1 (host header set to `lb-app-a.example.com`)
```
```suggestion
    - Origin 1 (`Host` header set to `lb-app-a.example.com`)
```

pedrosousa commented on products/load-balancing/src/content/understand-basics/monitors.md:
```diff
@@ -257,6 +257,44 @@ The sub-string must appear within the first 10KiB of your response body.
+  - Pool 2:
+
+    - Origin 3
+    - Origin 4 (host header set to `lb-app-b.example.com`)
```
```suggestion
    - Origin 4 (`Host` header set to `lb-app-b.example.com`)
```

pedrosousa commented on products/load-balancing/src/content/understand-basics/monitors.md:
```diff
@@ -257,6 +257,44 @@ The sub-string must appear within the first 10KiB of your response body.
+    - Origin 3
+    - Origin 4 (host header set to `lb-app-b.example.com`)
+
+- Monitor (host header set to `www.example.com`)
```
```suggestion
- Monitor (`Host` header set to `www.example.com`)
```

pedrosousa commented on products/load-balancing/src/content/understand-basics/monitors.md:
```diff
@@ -257,6 +257,44 @@ The sub-string must appear within the first 10KiB of your response body.
+
+- Monitor (host header set to `www.example.com`)
+
+In this scenario, health checks for **Origin 1** would use `lb-app-a.example.com`, health checks for **Origin 4** would use `lb-app-b.example.com`, and all other health checks would default to the `www.example.com`. For more on updating custom host configuration to be compatible with Cloudflare, see [Configure Cloudflare and Heroku over HTTPS](https://support.cloudflare.com/hc/articles/205893698) ...
```
A few suggestions:
- Drop the `the` in "to ~the~ `www.example.com`".
- Add `information` after "For more **information**".
- Add `your` to "updating custom host configuration": "updating **your** custom host configuration".

pedrosousa commented on products/load-balancing/src/content/understand-basics/monitors.md:
```diff
@@ -257,6 +257,44 @@ The sub-string must appear within the first 10KiB of your response body.
+
+1. On a monitor, select **Edit**.
+1. Select **Advanced health check settings**.
+1. If you have an origin overrides, you will see **Origin host header overrides**.
```
Should be either "If you have ~an~ origin overrides" or "If you have an origin override~s~" without the `s`.

pedrosousa commented on products/load-balancing/src/content/understand-basics/monitors.md:
```diff
@@ -257,6 +257,44 @@ The sub-string must appear within the first 10KiB of your response body.
+1. Select **Advanced health check settings**.
+1. If you have an origin overrides, you will see **Origin host header overrides**.
+
+![List of origin host header overrides](../static/images/origin_host_header_override.png)
```
I would use dashes in the image filename instead of underscores.

pedrosousa commented on products/load-balancing/src/content/understand-basics/monitors.md:
```diff
@@ -257,6 +257,44 @@ The sub-string must appear within the first 10KiB of your response body.
+
 ## Managing monitors via the Load Balancing dashboard
 
 Use the **Create Load Balancer** or **Edit Load Balancer** panels in the Load Balancing dashboard to manage health check monitors. For step-by-step guidance, see _[Create, attach, and configuring health checks](/create-load-balancer-ui#create-attach-and-configure-health-checks)_.
```
I'm never sure whether to format a link using italics... This one is up for discussion, or do send me a link about the criteria we use. :)

pedrosousa commented on products/load-balancing/src/content/understand-basics/pools.md:
```diff
@@ -215,6 +215,20 @@ Read only
+
+Since these examples require specific hostnames per origin, your load balancer could not properly route traffic _without_ a `Host` header override.
+
+If you need an origin `Host` header override, add it when [creating](/create-load-balancer-ui#create-and-add-origin-pools) or editing a pool. For security reasons, this header also needs to be a subdomain of the overall zone. See [Configure Cloudflare and Heroku](https://support.cloudflare.com/hc/articles/205893698-Configure-Cloudflare-and-Heroku-over-HTTPS) for more details.
```
```suggestion
If you need an origin `Host` header override, add it when [creating](/create-load-balancer-ui#create-and-add-origin-pools) or editing a pool. For security reasons, this header also needs to be a subdomain of the overall zone. See [Configure Cloudflare and Heroku](https://support.cloudflare.com/hc/articles/205893698) for more details.
```

pedrosousa commented on products/load-balancing/src/content/create-load-balancer-api/index.md:
```diff
@@ -243,7 +243,7 @@ If the response is an error, check the error message for a suggestion. If you’
 
 ### Create Pool 2
 
-To create a second pool, use the same command you did to create Pool 1, but give Pool 2 a different `name`.
```
Consider using backticks for `backup-server-4`.

pedrosousa commented on products/load-balancing/src/content/create-load-balancer-api/index.md:
```diff
@@ -261,6 +261,11 @@ To create a second pool, use the same command you did to create Pool 1, but give
       "name": "backup-server-4",
       "address": "0.0.4.0",
+      "header": {
+        "Host": [
```
The API accepts a list for the `Host` header? It's a bit strange... I'd expect it to be the other way around: `header` accepting a list, and `Host` having a single value. (Just a heads-up.)

kodster28 replied:
Interesting. I'll pass that along to the PM.
